### PR TITLE
Let deep subagents ride out a provider capacity outage

### PR DESCRIPTION
**Pull request by devxoul**
## Background

A PR auto-review posted "the reviewer subagent terminated twice without a verdict" and left the PR unreviewed. The reviewer had not misbehaved — the model provider was returning `overloaded_error` for about a minute. Both reviewer attempts made **3 model calls each and gave up in ~4 seconds**; the session 90 seconds later worked fine.

Two independent defects turned a transient capacity blip into a user-visible review failure.

**1. An overload gets no retry at all on a single-ref chain.**

`isRetryableSameRef` excludes throttle/overload on the stated premise that failover should handle it. But `promptWithFallback` is a documented no-op once there is no ref left to advance to — and `models.deep` is commonly one ref, so every ref is the last one. The two policies each defer to the other and nobody retries. Confirmed in production logs: zero `falling back` lines, because `onAttemptFailed` is never even reached on a length-1 chain.

**2. The required-block guard could not tell an outage from a bad answer.**

It only checked `hasRequiredBlock()`, so a turn that died upstream looked identical to a model that declined to emit `<review>`. It spent both recovery nudges re-prompting the same dead provider — each returning in ~1.4s — then installed a fallback blaming the subagent. Total budget burned: ~4s against a ~60s outage.

Worth noting the outage is invisible in logs: the upstream returns it as **HTTP 200 with a ~157-byte SSE error frame**, so the fetch observer records `status=200 error=null`.

## Changes

Add a `RetryPolicy` (`'responsive' | 'patient'`) to the turn runner.

- On the **last usable ref, and only there**, a patient caller retries an overload on a capacity-sized curve (10/20/30/30/30/30s, ±20% jitter, 180s window) instead of the 1–5s transport curve. Gating on the terminal ref matters: without it a 3-ref chain could wait ~9 minutes before exhausting failover.
- Earlier refs still fail over immediately. The activity fence is unchanged — a turn that already produced output or started a tool is still never replayed.
- `promptWithSameRefRetryOnly` now reports `{ success, lastError }` so the drain and nudge turns feed the same signal.
- The required-block guard tracks the latest turn outcome across the initial prompt, the drain, and the nudges. On an upstream failure it skips nudging entirely, stops early if a nudge dies the same way, and names the real cause in the fallback.

**Only `deep` subagents are patient**, per the latency requirement: the default profile exists to be responsive and a chat user will not wait. Keying off the **requested** profile (not the resolved one) keeps reviewer/researcher patient when `models.deep` is unconfigured, while an explicit `profileOverride` onto a responsive tier opts back out. Channel chat, the TUI, and cron pass no profile, so interactive latency is unchanged.

The verdict sentinel stays `incomplete-review-not-a-verdict` for both causes — the downstream contract is identical (no analysis landed), and a second sentinel would have to be taught to every policy branch in `typeclaw-channel-github/SKILL.md` that keys off the first.

## Testing

`bun run lint` / `format:check` / `typecheck` / `test` all pass (12044 pass, 0 fail).

New coverage:
- Patient caller rides out an outage and recovers; a responsive caller on the **same input** stops after 1 attempt (this pair is the mutation-proof that the new branch is doing the work).
- Omitting the policy preserves historical behavior.
- Retry budget and the 180s window both terminate the loop.
- A non-terminal ref fails over immediately instead of waiting.
- No patient replay after the model produced output.
- Provider failure spends **zero** nudges and renders `UPSTREAM PROVIDER FAILURE`; a model that merely omits the block still gets the full budget and the old wording; a nudge that dies upstream stops the loop.

## Review notes

- The 180s window bounds when the last retry may *start*, not total elapsed time — an in-flight attempt still runs to the observer's 300s deadline, so the pathological ceiling is ~8m, well inside the 30m reviewer/researcher spawn timeout.
- `patientBackoffMs` is a test seam, consistent with the existing `now` / `retryRandom` / `circuit` injection in the same file.
- The circuit breaker cannot self-trip from this: `recordThrottle` already fires once per ref after same-ref retries are exhausted, not per internal retry.
- `Retry-After` is not honored yet. The observer only logs the header today and the incident carried `retry_after=null`, so plumbing it through was not worth the cross-request state in this change.

### CODE REVIEW

File: src/agent/subagents.ts [hunk 1]
```diff
@@ -346,6 +356,7 @@ export async function invokeSubagent(name: string, options: InvokeSubagentOption
           refs: resolveFallbackChain(getConfig().models, resolveSubagentProfile(subagent, sessionOptions)),
```
[the remark this answers is not in the document]
Comment (devxoul) [reply src/agent/subagents.ts:373]: Fixed. Added a `recordHardProviderFailure` boundary in `runSession` that runs `detectHardProviderError` on a thrown error: a recognized provider failure is recorded as `latestTurnFailure` (so it reaches the required-block fallback and emits the sentinel), and anything unrecognized is rethrown untouched.

Wrapped all three turn sites — initial, drain, and nudge — since `promptWithSameRefRetryOnly` rethrows on the latter two as you noted. `throwIfLoopGuardAborted` stayed inside the initial try so abort ordering relative to the drain is unchanged; loop-guard aborts and spawn timeouts fall through `detectHardProviderError` as `null` and still propagate.

Three regression tests added: a thrown overload on the initial turn, a thrown overload on a nudge turn (stops the loop after one nudge), and a thrown `TypeError` asserting the subagent still fails rather than being masked as an outage.

File: src/agent/turn-runner.ts [hunk 2]
```diff
@@ -145,14 +163,36 @@ export async function promptPersistentTurnWithFallback(opts: {
         // Retry within budget only when the failure is same-ref retryable and
         // either no visible/tool activity occurred, or the caller explicitly
         // authorizes the strict post-tool-result resume recipe.
-        const retryableWithinBudget = retry < RETRIES_PER_REF && isRetryableSameRef(outcome.error.message)
+        // Overload means "this ref is out of capacity NOW", so the standing policy
+        // is to fail OVER rather than burn same-ref retries — which is why
+        // isRetryableSameRef excludes it. But on the LAST usable ref there is
+        // nothing to fail over to, and a single-ref chain is all last, so that
+        // policy leaves a capacity outage with no recovery whatsoever. A patient
+        // call site rides it out here instead; a responsive one still fails fast.
```
[the remark this answers is not in the document]
Comment (devxoul) [reply src/agent/turn-runner.ts:171]: Fixed in both paths. The backoff is now computed *before* the retry is accepted and the check is `elapsed + delay <= WINDOW` instead of `elapsed < WINDOW`, so a retry can never be scheduled to start past the advertised deadline. `now()` is sampled once per iteration and reused for both the elapsed calculation and the window-start latch.

Applied identically in `promptWithSameRefRetryOnly` so the two cannot drift.

Replaced the window test with a boundary case that advances a fake clock through the backoff: each attempt burns `WINDOW - 50ms`, so the second retry sits 50ms inside the window with a 100ms backoff pending. The old elapsed-only check accepted it; the new one rejects it, and the test asserts exactly 2 attempts.

File: src/agent/subagents.ts [hunk 3]
```diff
@@ -367,7 +389,10 @@ export async function invokeSubagent(name: string, options: InvokeSubagentOption
           },
         })
         if (result.success) activeModelRef = result.refUsed
+        latestTurnFailure = result.success ? undefined : result.lastError
         throwIfLoopGuardAborted(session)
```
[the remark this answers is not in the document]
Comment (devxoul) [reply src/agent/subagents.ts:399]: Good catch — that was a real regression. Fixed: `recordHardProviderFailure` now rethrows immediately when `requiredBlockTag` is undefined, so only reviewer and researcher absorb the throw. Everything else (explorer, operator, scout) keeps failing as before.

Moved the `requiredBlockTag` declaration above the helper rather than relying on closure-time TDZ resolution, so the dependency is explicit.

Added a regression test spawning `explorer` with a thrown overload and asserting `ok:false` — it fails without the guard.
